### PR TITLE
[FIX] base: prevent unlink groups

### PR DESCRIPTION
**Pull request by fmdl**
**Description of the issue/feature this PR addresses:**
With a wrong manipulation you can unlink a group.
During unlink 
- the `ir.model.access` is also unlink, 
- `ir.rule` are not unlink and become global

**Current behavior before PR:**
Go to runbot.
Unlink Portal groups
--> all Portal `ir.rule` become global --> there are lot of issue with access (you don't see all records).

**Desired behavior after PR is merged:**
You cannot unlink a group if it is used.

@odony 

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by fmdl** [comment]
OPW #2007628

**Comment by nim-odoo** [comment]
@Elkasitu @rco-odoo FYI

File: odoo/addons/base/models/res_users.py [hunk 1]
```diff
@@ -180,6 +180,13 @@ def write(self, vals):
         self.env['res.users'].has_group.clear_cache(self.env['res.users'])
         return super(Groups, self).write(vals)
 
+    def unlink(self):
+        if self.env['ir.rule'].sudo().search([('groups', 'in', self.ids)], limit=1) or\
+                self.env['ir.model.access'].sudo().search([('group_id', 'in', self.ids)], limit=1) or\
+                self.env['res.users'].sudo().search([('group_id', 'in', self.ids)], limit=1):
+            raise UserError(_('This group is used (Record Rule, Access Rights, Users), you can not delete it.'))
+        return super(Groups, self).unlink()
```
Comment (odony) [inline odoo/addons/base/models/res_users.py]: in P3 we can start using `super()` without parameters, it's more readable

**Comment by fmdl** [comment]
@odony thanks

File: odoo/addons/base/models/res_users.py [hunk 2]
```diff
@@ -180,6 +180,13 @@ def write(self, vals):
         self.env['res.users'].has_group.clear_cache(self.env['res.users'])
         return super(Groups, self).write(vals)
 
+    def unlink(self):
+        if self.env['ir.rule'].sudo().search([('groups', 'in', self.ids)], limit=1) or\
```
[the remark this answers is not in the document]
Comment (sswapnesh) [reply odoo/addons/base/models/res_users.py]: > This case is rare enough, and the number of rules low enough, optimizations (limit=1 or whatever) are not necessary.

 Hello @odony  
Could you please point out from which point` limit=1` helps to make search faster ? (Ref from https://github.com/odoo/odoo/pull/30003#issuecomment-452605666)

File: odoo/addons/base/models/res_users.py [hunk 3]
```diff
@@ -180,6 +180,13 @@ def write(self, vals):
...
+                self.env['ir.model.access'].sudo().search([('group_id', 'in', self.ids)], limit=1) or\
+                self.env['res.users'].sudo().search([('group_id', 'in', self.ids)], limit=1):
+            raise UserError(_('This group is used (Record Rule, Access Rights, Users), you can not delete it.'))
```
[the remark this answers is not in the document]
Comment (sswapnesh) [reply odoo/addons/base/models/res_users.py]: Better to have
`Please deactivate or delete linked record rules before deleting their group.`

**Comment by fmdl** [comment]
ping @nim-odoo

**Comment by odony** [comment]
@fmdl why re-target to 11.0? It seems more a "nice-to-have" than a bugfix to me, so I'd go for master. 

Besides if we target 11.0 the `super()` call needs to be written in P2 fashion, as 11.0 is supposed to have dual P2-P3 compatibility.

**Comment by fmdl** [comment]
@odony, it is moved in master.

**Comment by fmdl** [comment]
ping @odony

**Comment by fmdl** [comment]
ping @odony

**Comment by fmdl** [comment]
close by https://github.com/odoo/odoo/pull/48544

**Pull request closed without merging**